### ISSUE TITLE
Experiment with different function combinations and exp setups

### ISSUE THREAD

**Issue by stasinos** [stasinos/machine_learning#13]
Have a look at exp01. This experiment defines the network fn1(x) + fn2(2x), where fn1 and fn2 are different instances of our "universal" approximator. The network is trained on data generated from sin(x) + sin(2x).

Inspect visually the test plot for fn1, fn2, and the overall network. fn1 and fn2 vaguely resemble sin(x) and sin(2x), so although far from perfect, they are promising (See below, but also generate them to ensure we all get the same results)

Experiment with different function combinations AND combinators (composition, addition) to draw conclusions about what works and what doesn't. Look for the hyperparams and training set size that seem to work best, but must work for all combinations. (That is, find a single set of hyperparams that works best for all complex functions, not a different set of hyperparams for each complex function)

sin(x) + sin(2x):

![sine_on_sine](https://github.com/stasinos/machine_learning/assets/663585/da7c5ce2-6599-44ff-9f91-c6041afa15eb)

fn1(x) + fn2(2x):

![exp01_final](https://github.com/stasinos/machine_learning/assets/663585/e41aabd7-bd2c-47c3-9615-d153860b57f7)

fn1(x):

![exp01_model1](https://github.com/stasinos/machine_learning/assets/663585/3378139e-de83-498a-b752-2c4140693693)

fn2(x):

![exp01_model2](https://github.com/stasinos/machine_learning/assets/663585/566d57c5-f661-4eb3-b069-ae260ef863fa)

**Comment by stasinos** [comment]
It was thinking that maybe if the learner is presented with different domain ranges, it might be able to figure out that fn1 and fn2 are periodic.

**Comment by stasinos** [comment]
Just realized that I had overlooked telling the learner that sin(x)+sin(2x) combines two instances of the same function. So exp02 makes this information available. The results again look promising, but not really there.

Plot of fn(x)+fn(2x):

![exp02_final](https://github.com/stasinos/machine_learning/assets/663585/c65608bc-9bfd-4ab3-bf4c-556e859c572e)

Plot of fn(x):

![exp02_model0](https://github.com/stasinos/machine_learning/assets/663585/6532bd73-ea84-4a8c-a3cc-49ba40d0c593)

**Comment by stasinos** [comment]
The branch for this issue should be from autodiff, not from master

**Comment by tatiana-boura** [comment]
> The branch for this issue should be from autodiff, not from master

Sorry, I am creating it from the Github interface and I guess it is automatically forked from master. Will be more careful in the future.

Also, shouldn't the function ```sine_in_01(x)``` be
```Python
def sine_in_01(x):
    xx = torch.sin(x) + torch.sin(2*x)
    return 0.25*( 2 + xx )
```
instead of
```Python
def sine_in_01(x):
    xx = torch.sin(x) + torch.sin(2*x)
    return 0.5*( 1 + xx )
```
and ```nnrange_to_realrange(x)``` :
```Python
def nnrange_to_realrange(x):
    return 4*x-2
```
instead of
```Python
def nnrange_to_realrange(x):
    return 2*x-1
```
?

**Comment by stasinos** [comment]
Good catch, please fix and commit.

**Comment by stasinos** [comment]
> I am creating it from the Github interface and I guess it is automatically forked from master.

No worries, it's easy to fix, at least from the cmd line. Just merge autodiff-exercise into the new branch.

**Comment by tatiana-boura** [comment]
Actually, I was incorrect before. There is a minor conceptual bug. The output of the model (```SingleSinTwice```) is bound in $0 \leq sin(x)+sin(2x) \leq 2$, since $0 \leq f(x) \leq 1$ and $0 \leq f(2x) \leq 1$ . So, we either bound the target values in $[0,2]$ or bound the output of the model from  $[0,2]$ to $[0,1]$. 

If we bound the target values in $[0,2]$ by changing ```sine_in_01(x)``` and ```nnrange_to_realrange(x):``` to

```Python
def sine_in_02(x):
    xx = torch.sin(x) + torch.sin(2*x)
    return 0.5*(2 + xx)

def nnrange_to_realrange(x):
    return 2*x-2
```

and add a Sigmoid layer to the ```FunctionApproximator``` at the very end, the results are kinda better:

Plot of fn(x)+fn(2x):
![exp01b_final](https://github.com/stasinos/machine_learning/assets/61378350/78ce0f02-68b7-4342-a740-e36dd5441126)

Plot of fn(x):
![exp01b_model0](https://github.com/stasinos/machine_learning/assets/61378350/0fcd4745-e4a5-4e98-b37f-7b54915bf15f)

I also tried bounding the  output of the model using a final Sigmoid layer in ``SingleSinTwice``, but the results were not that good as with the above approach.

Anyway, I am still trying things, so I will let you know if there are any shocking results.

**Comment by tatiana-boura** [comment]
So, if we replace ```Tanh``` with the ``Snake`` activation function  ([https://proceedings.neurips.cc/paper/2020/file/1160453108d3e537255e9f7b931f4e90-Paper.pdf](url) / https://pypi.org/project/torch-snake/ ) the results for the $sin(x)+sin(2\cdot x)$ example are the following :

Plot of fn(x)+fn(2x):

![exp01b_final](https://github.com/stasinos/machine_learning/assets/61378350/7d4fc6b6-3310-4009-9b27-ba62738f212c)

Plot of fn(x) (it is for some reason defined in a wrong range, I guess due to the rearrangement, but will look into it):

![exp01b_model0](https://github.com/stasinos/machine_learning/assets/61378350/18fe8143-dac8-46c7-98e1-2f5defb867cf)


Fingers crossed it works well for the other, non-periodic functions in our disposal.

**Comment by tatiana-boura** [comment]
So, if we replace ```Tanh``` with ```Snake``` activation function ([https://proceedings.neurips.cc/paper/2020/file/1160453108d3e537255e9f7b931f4e90-Paper.pdf](url) / [https://pypi.org/project/torch-snake/](url)) the results are the following for the $sin(x)+sin(2x)$ example:

Plot of fn(x)+fn(2x):

![exp01b_final](https://github.com/stasinos/machine_learning/assets/61378350/c6d30c9d-a3fa-447e-9360-5b8dab96f1fc)

Plot of fn(x):

![exp01b_model0](https://github.com/stasinos/machine_learning/assets/61378350/d6c73f3a-9450-46ae-9636-83e3d19bfa2e)

Fingers crossed that is works equally well with the other, non-periodic functions in our disposal.

**Comment by stasinos** [comment]
That's impressive, but as you also note, applying it might require prior knowledge of which functions are periodic; Or a way to decide which approximation is better after trying both snake and tanh. You might want to re-open Issue https://github.com/stasinos/machine_learning/issues/6 and repeat the experiments on the individual elementary functions, before deciding how to proceed.

**Comment by stasinos** [comment]
Well done BTW, great work.

**Comment by tatiana-boura** [comment]
An update. The network architecture with the ```Snake``` activation functions worked perfectly for functions $sin(\alpha x) + sin(\beta x)$. However, it failed to approximate $f(x)$, $sin(x)$ and  $f_2(x)$ for $f(x) = sin(f_2(x))$. 

I tried, again, using ```Tanh``` and the results where better. Take a look.

With ```Tanh```:

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/ce3da839-f663-425d-b81d-ccab4d2ee350" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/0643dd69-b6a4-46f7-9fe5-fd20ecb76984" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/e09f1c00-eae1-4b61-b6ce-c9eedf8f882c" width="250" /> 

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/66756f96-0945-47a1-a340-08ee70fca9c7" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/62a5b6c6-ec04-436f-98d9-51c17721e154" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/1aafad35-933f-41ed-bdd6-9368796524cd" width="250" /> 

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/e03678c4-1716-4fd5-ac19-a3df1a974559" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/a1478c16-2fa9-414a-9ab9-b2e508e28fa3" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/6021af29-956f-46d1-8cff-5da565980810" width="250" /> 

With ```Snake```:

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/a90bdbb3-0099-4bef-bc45-394493728997" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/c29a544d-ae9e-49fe-ad3b-dfbdcc3041b1" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/71fcb37b-c0aa-4226-ae82-75a59b61ea37" width="250" /> 

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/d818ad8d-a187-4cf6-b546-fa65fba0f196" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/439491e8-6992-46f5-ad32-66a0a2203dba" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/206a6dd3-37b6-4614-bc9d-f95cbaed9850" width="250" /> 

<p float="middle">
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/7bfdeb73-9ae0-4069-9544-06a7b11459ae" width="250" />
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/54db3803-8cc0-407f-a321-7bff9421ada1" width="250" /> 
  <img src="https://github.com/stasinos/machine_learning/assets/61378350/1af887df-7536-4a2e-bd06-d01e44cb4419" width="250" />

**Comment by stasinos** [comment]
Some of these are compositions, if I understood correctly. Have you changed the forward() function accordingly?

**Comment by tatiana-boura** [comment]
> Some of these are compositions, if I understood correctly. Have you changed the forward() function accordingly?

Yes of course. Will try again after the changes from #14 .

**Comment by stasinos** [comment]
We should start wrapping this up. The experimental setup seems enough of an infrastructure for the experiments you need to show for this exercise.

Not everything needs to work perfectly to close this issue, as long as you know what doesn't work for methodological reasons and what might be a bug and should not be presented.

Comparing actual plots will have to get out of scope, this was harder than expected and the current setup gives up enough to write an interesting report.

I thank you for the patience and help, carrying out an exercise while it was being formulated. I hope you enjoyed the journey as much as I did.

**Comment by tatiana-boura** [comment]
> We should start wrapping this up. The experimental setup seems enough of an infrastructure for the experiments you need to show for this exercise.

Okay! Will wrap up the experiments, push the corresponding scripts and call it a day! When I make the final commit for this issue I will close it.

> I thank you for the patience and help, carrying out an exercise while it was being formulated. I hope you enjoyed the journey as much as I did.

It was very fun! Thank you for the experience.

**Comment by KChald** [comment]
It was really interesting! Thank you for this!

**Comment by KChald** [comment]
I had some revelations that would be interesting to share before closing. Hope you enjoy!

 With a smaller learning rate of 1e-04, snake activation function, a dataset size of 64x1024 and 60< epochs ( ~3 minutes due to small dataset) the following results were obtained:

![1](https://github.com/stasinos/machine_learning/assets/102811531/5197c39a-5879-408f-a4a6-2ec8edb715e7)

![expsinc](https://github.com/stasinos/machine_learning/assets/102811531/ab617617-927e-45a1-afc5-1b0963112647)

These parameters were also tested in other functions, yielding similar results for the main function.

Also, I would like to propose an explanation on why the sin(exp(x)) is not learnable in (-60,60) . As we can see in the range (-60,60) the output looks like this.

![6](https://github.com/stasinos/machine_learning/assets/102811531/3a4ad7d1-34d4-4261-85fd-88ba0c1a414f)

It is obvious not possible for our model to learn a "square" like output due to limited amount of inputs. This can also be supported by the fact that when we use (-10,10) range and sin(20exp(x)) the model slowly converges . Also observe the exponential time the model needs to learn the periods as they come closer and closer.
With same parameters but more epochs:

After 10 epochs:

![exp05_10](https://github.com/stasinos/machine_learning/assets/102811531/eaeffec2-f688-4b2f-9317-ad59a1549151)

After 150 epochs:

![exp05_150](https://github.com/stasinos/machine_learning/assets/102811531/f873f52d-f395-4372-b94c-28e2b97126c9)

After 400 epochs

![exp05_final](https://github.com/stasinos/machine_learning/assets/102811531/b476bfb2-a702-4f98-8707-9165c99fd623)

The model continues to yield results as the epochs increase but we need an exponential amount of time to converge, if even possible after a number of epochs. 

Last but not least, something that I found interesting and could be used as an example of learning 2 functions that compose the main, without learning the ones we want the model to learn, is below where it is obvious that x=0 and y = x are combined to produce the exp(lnx).

![2](https://github.com/stasinos/machine_learning/assets/102811531/b32efbf4-5932-4bda-bfdd-576037e6d7f9)

![4](https://github.com/stasinos/machine_learning/assets/102811531/0f83e7db-c59c-4f5b-a213-9a3a469c69f8)

![3](https://github.com/stasinos/machine_learning/assets/102811531/40a31dd3-1776-4c87-a129-8f97461c896b)

Also, it reveals that the axes might have a problem since y = x = exp(ln(x)) and we can observe in the graph that y ≈ x/5.

Have a good day!

**Comment by stasinos** [comment]
Nice catches.

sed s/"square" like output/practically inconsistent output/

It would take the ability to extrapolate from the parts of the domain that are clearer, rather than approximate. You might want to mention in your conclusions as possible future work, that a clever manipulation of the underlying NNs at the autodiff layer can recognize periodicity and train one periodic and one non-periodic NN.
